federatedscope/vertical_fl/trainer/feature_order_protected_trainer.py

```diff
@@ -25,44 +25,88 @@ def get_feature_value(self, feature_idx, value_idx):
-                idx_start = j * bin_size
-                idx_end = min((j + 1) * bin_size, len(raw_feature_order[i]))
-                feature_order_frame = raw_feature_order[i][idx_start:idx_end]
-                np.random.shuffle(feature_order_frame)
-                _protected_feature_order.append(feature_order_frame)
-                if self.share_bin:
-                    if j != self.bin_num - 1:
-                        split_position[i].append(idx_end)
-                    min_value = min(data[feature_order_frame][:, i])
-                    max_value = max(data[feature_order_frame][:, i])
-                    if j == 0:
-                        self.split_value[i][idx_end] = max_value
-                    elif j == self.bin_num - 1:
-                        self.split_value[i][idx_start] += min_value / 2.0
+        bucket_size = int(
+            np.ceil(self.cfg.dataloader.batch_size / self.bucket_num))
+        if self.epsilon is None:
+            prob_for_preserving = 1.0
+        else:
+            _tmp = np.power(np.e, self.epsilon)
+            prob_for_preserving = _tmp / (_tmp + self.bucket_num - 1)
+        prob_for_moving = (1.0 - prob_for_preserving) / (self.bucket_num - 1)
+        split_position = []
+        self.split_value = []
+
+        for feature_idx in range(len(raw_feature_order)):
+            bucketized_feature_order = self._bucketize(
+                raw_feature_order[feature_idx], bucket_size, self.bucket_num)
+            noisy_bucketizd_feature_order = [[]
+                                             for _ in range(self.bucket_num)]
+
+            # Add noise to bucketized feature order
+            for bucket_idx in range(self.bucket_num):
+                probs = np.ones(self.bucket_num) * prob_for_moving
+                probs[bucket_idx] = prob_for_preserving
+                for each in bucketized_feature_order[bucket_idx]:
+                    selected_bucket_idx = np.random.choice(list(
+                        range(self.bucket_num)),
+                                                           p=probs)
+                    noisy_bucketizd_feature_order[selected_bucket_idx].append(
+                        each)
+
+            # Save split positions (instance number within buckets)
+            # We exclude the endpoints to avoid empty sub-trees
+            _split_position = list()
+            _split_value = dict()
+            accumu_num = 0
+            for bucket_idx, each_bucket in enumerate(
+                    noisy_bucketizd_feature_order):
+                instance_num = len(each_bucket)
+                # Skip the empty bucket
+                if instance_num != 0:
+                    # Skip the endpoints
+                    if bucket_idx != self.bucket_num - 1:
+                        _split_position.append(accumu_num + instance_num)
+
+                    # Save split values: average of min value of (j-1)-th
+                    # bucket and max value of j-th bucket
```

Comment:
The min and max should be switched from this line to line 94, since the order is increasing.